renderdoc/driver/vulkan/vk_counters.cpp

```diff
@@ -376,6 +514,197 @@ std::vector<CounterResult> VulkanReplay::FetchCountersAMD(const std::vector<GPUC
   return ret;
 }
 
+struct VulkanKHRCallback : public VulkanDrawcallCallback
+{
+  VulkanKHRCallback(WrappedVulkan *vk, VulkanReplay *rp, VkQueryPool qp)
+    : m_pDriver(vk),
+      m_pReplay(rp),
+      m_QueryPool(qp)
+  {
+    m_pDriver->SetDrawcallCB(this);
+  }
+  ~VulkanKHRCallback() { m_pDriver->SetDrawcallCB(NULL); }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd) override
+  {
+    ObjDisp(cmd)->CmdBeginQuery(Unwrap(cmd), m_QueryPool, (uint32_t)m_Results.size(), 0);
+  }
+
+  bool PostDraw(uint32_t eid, VkCommandBuffer cmd) override
+  {
+    ObjDisp(cmd)->CmdEndQuery(Unwrap(cmd), m_QueryPool, (uint32_t)m_Results.size());
+    m_Results.push_back(eid);
+    return false;
+  }
+
+  void PostRedraw(uint32_t eid, VkCommandBuffer cmd) override {}
+  // we don't need to distinguish, call the Draw functions
+  void PreDispatch(uint32_t eid, VkCommandBuffer cmd) override { PreDraw(eid, cmd); }
+  bool PostDispatch(uint32_t eid, VkCommandBuffer cmd) override { return PostDraw(eid, cmd); }
+  void PostRedispatch(uint32_t eid, VkCommandBuffer cmd) override { PostRedraw(eid, cmd); }
+  void PreMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override { PreDraw(eid, cmd); }
+  bool PostMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override
+  {
+    return PostDraw(eid, cmd);
+  }
+  void PostRemisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override
+  {
+    PostRedraw(eid, cmd);
+  }
+  void AliasEvent(uint32_t primary, uint32_t alias) override
+  {
+    m_AliasEvents.push_back(std::make_pair(primary, alias));
+  }
+
+  void PreEndCommandBuffer(VkCommandBuffer cmd) override {}
+  WrappedVulkan *m_pDriver;
+  VulkanReplay *m_pReplay;
+  VkQueryPool m_QueryPool;
+  std::vector<uint32_t> m_Results;
+  // events which are the 'same' from being the same command buffer resubmitted
+  // multiple times in the frame. We will only get the full callback when we're
+  // recording the command buffer, and will be given the first EID. After that
+  // we'll just be told which other EIDs alias this event.
+  std::vector<std::pair<uint32_t, uint32_t> > m_AliasEvents;
+};
+
+std::vector<CounterResult> VulkanReplay::FetchCountersKHR(const std::vector<GPUCounter> &counters)
+{
+  std::vector<uint32_t> counterIndices;
+  for (const auto& c : counters)
```

Comment:
I prefer concrete types over use of `auto` unless the type is particularly messy, like lambdas and iterators.

Reply:
Agreed.